docs/source/torch.compiler_dynamo_deepdive.rst

```diff
@@ -0,0 +1,873 @@
+graph is only valid if these assumptions hold on the new inputs.
+
+For example, any constant input to a function, like a string, installs a
+guard stating that that input should be of type ``str`` and equal to the
+string we passed. Running
+
+.. code:: python!
+
+   import torch
+
+   @torch.compile
+   def fn(a, b):
+       return a * len(b)
+
+   fn(torch.arange(10), "Hello")
+
+with ``TORCH_LOGS=guards`` prints (among other guards)
+
+.. code:: python!
+
+   ___check_type_id(L['b'], 94334122025024)
+   L['b'] == 'Hello'
+
+This reads as “the local variable ``b`` should have a specific type
+(``str`` in this case) and its value should be ``'Hello'``”. If we then
+execute the function again passing a different argument
+
+.. code:: python!
+
+   import torch
+
+   @torch.compile
+   def fn(a, b):
+       return a * len(b)
+
+   fn(torch.arange(10), "Hello")
+   fn(torch.arange(10), "Hi")
+
+we can see the guard that failed by running ``TORCH_LOGS=recompiles``
+
+.. code:: python!
+
+   Recompiling function fn in script.py:3
+   triggered by the following guard failure(s):
+        - L['b'] == 'Hello'
+
+Guards are accumulated while `the inputs to the function are wrapped in
+the
+builder <https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/variables/builder.py#L808-L810>`__
+and `during the execution of the
+program <https://github.com/pytorch/pytorch/blob/69f112d5867f785a3a090a0c6d6644ae047033ac/torch/_dynamo/variables/dicts.py#L763-L769>`__.
+We will show many more examples of guards in the next section, but first
+let us discuss sources.
+
+A **source** tracks how to reconstruct a variable from the original
+local or global variables present when entering the current frame. In
+particular, it tracks the original local and global objects and any of
+the objects they contain. In
+
+.. code:: python!
+
+   def foo(x: Tensor, y: List[Tensor]):
+       a = x * y[0]
+       return a * x
+
+``x`` and ``y`` have
+```LocalSource`` <https://github.com/pytorch/pytorch/blob/40dc0580a69565b06ec5263efe5d87cecc8200f7/torch/_dynamo/source.py#L80-L92>`__
+as their source, and ``y[0]`` has
+```GetItemSource`` <https://github.com/pytorch/pytorch/blob/40dc0580a69565b06ec5263efe5d87cecc8200f7/torch/_dynamo/source.py#L302>`__,
+which stores a ``LocalSource`` inside. On the other hand, ``a`` will not
+have a source as it is an intermediate variable that only exists within
+the fx graph.
+
+All these are defined in
+```torch/_dynamo/source.py`` <https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/source.py>`__.
+We can see the guard generated by ``GetItemSource`` in the following
+example:
+
+.. code:: python!
+
+   import torch
+
+   @torch.compile
+   def fn(x, l):
+       return x * len(l[0])
+
+   fn(torch.randn(8), ["Hi", "Hello"])
+
+generates the following guards
+
+.. code:: python!
+
+   ___check_type_id(L['l'], 94439025877664)
+   len(L['l']) == 2
+   hasattr(L['x'], '_dynamo_dynamic_indices') == False
```

Comment:
worth elaborating in a few more lines what this is

Reply:
Just removed that line to avoid distracting the reader.